woocommerce/woocommerce-ios
# Fix infinite scroll for grouped product's linked products after deletion

Fix infinite scroll for grouped product's linked products #2738

- [x] ⚠️ Make sure PR base is `develop` after https://github.com/woocommerce/woocommerce-ios/pull/2747 is merged ⚠️ 

## Changes

- Created `PaginationTracker` and `PaginationTrackerDelegate` that is an updated version of `SyncingCoordinator` with major differences:
  - No more caching behavior and syncing an earlier page. I don't feel like we need to sync an earlier page when scrolling to it, because if an earlier page has changes, we probably want to fetch the other earlier pages anyway. Our API doesn't work great with syncing, letting the client know of the changes. I think just keeping fetching the next highest page would be sufficient, and the user can still sync the first page by pulling down to refresh
  - The sync completion block in the delegate now takes `Result<Bool, Error>` instead of just a success boolean. This allows the pagination tracker to know if the API results have reached the end from the delegate, rather than the logic to compare the table view maximum index path with the page size.
- Updated `PaginatedListSelectorDataSource` and `PaginatedListSelectorViewController` to replace `SyncingCoordinator` with `ScrollWatcher` and `PaginationTracker`.
- Updated `PaginatedListSel` implementations to set `hasNextPage` on sync completion: `GroupedProductListSelectorDataSource`, `ProductListMultiSelectorDataSource`, `PaginatedProductShippingClassListSelectorDataSource`, and `ProductTaxClassListSelectorDataSource` (always no next page)

## Testing

Prerequisite: the store has a grouped product with more than 25 linked products
- Go to the Products tab
- Tap on the grouped product that has more than 25 linked products
- Tap on the "Grouped products" row
- Tap "X" to delete a linked product from the list
- Scroll down toward the end of the list --> the infinite scroll should work (this wouldn't work on `develop`)

Please also sanity check on other affected screens (now infinite scroll is triggered when scrolling to ~70% of the list):

### Product shipping classes

- Go to the Products tab
- Tap on a simple physical product
- Tap on the shipping row or from the bottom sheet, and then the shipping class row --> the paginated list should load as before, with the correct selection

### Product tax classes

- Go to the Products tab
- Tap on a simple product
- Tap on the price row, and then the tax class row --> the list should load as before, with the correct selection

### Grouped products

- Make sure the products feature switch is on under Settings > Experimental Features
- Go to the Products tab
- Tap on a grouped product
- Tap on the grouped products row --> the paginated list should load as before
- Tap "Add Products" CTA --> the paginated list of products should load as before

Update release notes:
